### PR TITLE
Graph generators for testing

### PR DESCRIPTION
**Pull request by mzargham**
Introduce simple graph generators as research infrastructure [WIP]

Graph generators are critical for:
- writing test cases for the python page_ranker implementation
- for exploring the way any particular page_ranker parameterization spreads cred

Sample graphs included:
- line graph
- circle graph
- star graph
- tree graph

Joint work with @BrianLitwin 

Test plan:
Tests added; require further review.

### CODE REVIEW

File: infra/graph_generators.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Sun Mar 17 13:00:33 2019
+
+@author: Zargham
```
Comment (teamdandelion) [inline infra/graph_generators.py:6]: What is the function of the author block? Does it represent "ownership" of the file in an abstract sense? What information does it provide outside of the information readily available from git blame? Are users who make modifications to this file expected to add themselves as authors?

Similarly, why do we have a created timestamp? In the few cases where the creation date is relevant information, it can be found by checking the timestamp on the associated commit. I don't see why humans should be manually maintaining such metadata.

If we do standardize on having explicit authorship tags, it should be the users' github handle, and not their last name.

File: infra/graph_generators.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+"""
+import networkx as nx
+
+def lineGraphGen(N, bidir=False,nodeTypeName='vanilla',edgeTypeName='vanilla'):
```
Comment (teamdandelion) [inline infra/graph_generators.py:10]: nit, python style for method names is underscore naming, as in `line_graph_gen`

Comment (teamdandelion) [inline infra/graph_generators.py:10]: nit: missing commas between spaces. run `black` on this code.

Comment (teamdandelion) [inline infra/graph_generators.py:10]: nit, arguments should have underscore naming, as in `edge_type_name`

Comment (teamdandelion) [inline infra/graph_generators.py:10]: why is edge_type_name and node_type_name needed?

Comment (teamdandelion) [inline infra/graph_generators.py:10]: `num_nodes` is a clearer parameter name than `N`.

File: infra/graph_generators.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    line = nx.path_graph(N, create_using=nx.MultiDiGraph)
```
Comment (teamdandelion) [inline infra/graph_generators.py:12]: this algorithm is a little funky. Given that we create the path_graph using nx.MultiDiGraph in either case, couldnt we do something like:

```python
graph = nx.path_graph(N, create_using=nx.MultiDiGraph)
if bidir:
  for e in edges:
    graph.add_edge(e[1], e[0])

# ... stuff ....
return graph
```

File: infra/graph_generators.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    if not(bidir):
+        G = line
```
Comment (teamdandelion) [inline infra/graph_generators.py:14]: nit: `G` is a bad variable name, especially as Capitalization in python implies a class name.

File: infra/graph_generators.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        for e in edges:
+            G.add_edge(e[0],e[1])
+            G.add_edge(e[1],e[0])
+    
+    nx.set_node_attributes(G,nodeTypeName, 'type')
+    nx.set_edge_attributes(G,edgeTypeName, 'type')
+    
+    return G
+
+def starGraphGen(N, kind='sink',nodeTypeName='vanilla',edgeTypeName='vanilla'):
```
Comment (teamdandelion) [inline infra/graph_generators.py:27]: nit: methods should have underscore names., `N` is a bad argument name

File: infra/graph_generators.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    
+    star = nx.star_graph(N)
+    G = nx.MultiDiGraph()
+
```
Comment (teamdandelion) [inline infra/graph_generators.py:31]: should have error handling that checks that `kind` is an accepted argument:

```python
if kind is not "source" and kind is not "sink" and kind is not "bidir":
  raise ValueError("Unrecognized kind: {}".format(kind)")
```

This allows much greater confidence that code using this API is calling it correctly.

File: infra/graph_generators.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            G.add_edge(e[0],e[1])
+        if (kind == 'sink') or (kind == 'bidir'):
+            G.add_edge(e[1],e[0])
+
+    nx.set_node_attributes(G,nodeTypeName, 'type')
+    nx.set_edge_attributes(G,edgeTypeName, 'type')
+    
+    return G
+
+def circleGraphGen(N, bidir=False,nodeTypeName='vanilla',edgeTypeName='vanilla' ):
```
Comment (teamdandelion) [inline infra/graph_generators.py:43]: same nits as before

File: infra/graph_generators.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    
+    circle = nx.cycle_graph(N, create_using=nx.MultiDiGraph)
```
Comment (teamdandelion) [inline infra/graph_generators.py:45]: same as before, we should not need to initialize two graphs in the bidir case, we should be able to make the graph bidirectional in-place.

File: infra/graph_generators.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        for e in edges:
+            G.add_edge(e[0],e[1])
+            G.add_edge(e[1],e[0])
+            
+    nx.set_node_attributes(G,nodeTypeName, 'type')
+    nx.set_edge_attributes(G,edgeTypeName, 'type')
+    
+    return G
+
+def treeGraphGen(r,h, kind='sink',nodeTypeName='vanilla',edgeTypeName='vanilla'):
```
Comment (teamdandelion) [inline infra/graph_generators.py:60]: same nits as on other methods

File: infra/graph_generators.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        G = tree
+    elif kind =='sink':
+        G = nx.MultiDiGraph()
+        for e in tree.edges:
+            G.add_edge(e[1],e[0])
+    elif kind == 'bidir':
+        G = nx.MultiDiGraph()
+        for e in tree.edges:
+            G.add_edge(e[1],e[0])
+            G.add_edge(e[0],e[1])
```
Comment (teamdandelion) [inline infra/graph_generators.py:74]: should add an else block to catch the case where kind is unrecognized:

```python
else:
  raise ValueError("unrecognized kind: {}".format(kind)
```

File: infra/graph_generators.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    nx.set_edge_attributes(G,edgeTypeName, 'type')
+    
+    return G
+
+def starGraphGen(N, kind='sink',nodeTypeName='vanilla',edgeTypeName='vanilla'):
+    
+    star = nx.star_graph(N)
+    G = nx.MultiDiGraph()
+
+    for e in star.edges:
```
Comment (teamdandelion) [inline infra/graph_generators.py:32]: since we only ever want that star graph to iterate over its edges, i think it would be cleaner to inline it:

```python
for e in nx.star_graph(N).edges:
  # do the stuff
```

Comment (teamdandelion) [reply infra/graph_generators.py:32]: for even more readability, maybe destructure it?
```python
for (src, dst) in nx.star_graph(N).edges:
  # do something
```

File: infra/graph_generators.py (hunk 8 above)
Comment (teamdandelion) [reply infra/graph_generators.py:45]: could think about factoring out a `make_bidirectional` method if we find ourselves doing this a lot.

File: infra/graph_generators.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    elif kind == 'bidir':
+        G = nx.MultiDiGraph()
+        for e in tree.edges:
+            G.add_edge(e[1],e[0])
+            G.add_edge(e[0],e[1])
+
+    nx.set_node_attributes(G,nodeTypeName, 'type')
+    nx.set_edge_attributes(G,edgeTypeName, 'type')
+    
+    return G
```
Comment (teamdandelion) [inline infra/graph_generators.py:79]: nit: trailing whitespace--hopefully `black` will take care of this.

File: infra/graph_generators.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    if not(bidir):
+        G = line
+    else:  
+        edges = line.edges
+        G = nx.MultiDiGraph()
+        for e in edges:
+            G.add_edge(e[0],e[1])
+            G.add_edge(e[1],e[0])
+    
+    nx.set_node_attributes(G,nodeTypeName, 'type')
```
Comment (teamdandelion) [inline infra/graph_generators.py:22]: if we do keep the nodeTypeName and edgeTypeName logic and arguments, we shouldn't keep repeating this logic all over the place. instead, we should have a `GraphOptions` type which includes a node type name and edge type name, and then a method like `_apply_graph_options(graph, options)` which does these mutations through a single re-used code path. then we can test it more thoroughly, easily change the logic, etc.

File: infra/graph_generators_test.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+class GraphGenerators_LineGraph(unittest.TestCase):
+    def test_lineGraphGen_sanity(self):
+        g = lineGraphGen(4)
+        self.assertEqual(list(g.edges()), [(0, 1), (1, 2), (2, 3)])
+
+    def test_lineGraphGen_bidir(self):
+        g = lineGraphGen(4, bidir=True)
+        # (node, neighbor)
+        self.assertEqual(list(g.edges()), [(0, 1), (1, 0), (1, 2), (2, 1), (2, 3), (3, 2)])
```
Comment (teamdandelion) [inline infra/graph_generators_test.py:20]: we should test the edge case too. what if we pass 0 to the graph generator?

File: infra/graph_generators_test.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+class GraphGenerators_StarGraph(unittest.TestCase):
+
+# how to validate that 'kind' is either 'source', 'sink', or 'bidir' ??
```
Comment (teamdandelion) [inline infra/graph_generators_test.py:24]: we should have the fn raise a ValueError if the kind isnt an acceptable value, then test that behavior here.

File: infra/graph_generators_test.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        # (node, neighbor)
+        self.assertEqual(list(g.edges()), [(0, 1), (1, 0), (1, 2), (2, 1), (2, 3), (3, 2)])
+
+class GraphGenerators_StarGraph(unittest.TestCase):
+
+# how to validate that 'kind' is either 'source', 'sink', or 'bidir' ?? 
+
+    def test_starGraph_sink(self):
+        g = starGraphGen(4)
+        self.assertEqual(list(g.edges()), [(1, 0), (2, 0), (3, 0), (4, 0)])
```
Comment (teamdandelion) [inline infra/graph_generators_test.py:28]: @BrianLitwin when we were chatting i mentioned i didn't like these test cases, but now that i am reading and paying closer attention, i think they're fine. its pretty easy to read and reason about what this graph is, and whether it's the right graph for the arguments.

File: infra/graph_generators_test.py [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+    def test_starGraph_source(self):
+        g = starGraphGen(4, kind='source')
+        self.assertEqual(list(g.edges()), [(0, 1), (0, 2), (0, 3), (0, 4)])
+
+    def test_starGraph_bidir(self):
+        g = starGraphGen(4, kind='bidir')
+        self.assertEqual(list(g.edges()), [(0, 1), (0, 2), (0, 3), (0, 4), (1, 0), (2, 0), (3, 0), (4, 0)])
```
Comment (teamdandelion) [inline infra/graph_generators_test.py:36]: this is kind of a doozy to keep all in my head--maybe check for `n=2` or `n=1` instead.

File: infra/graph_generators.py (hunk 2 above)
Comment (teamdandelion) [inline infra/graph_generators.py:10]: nit, all of these public methods should have python docstrings:
https://www.python.org/dev/peps/pep-0257/

File: infra/graph_generators_test.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+
+    def test_treeGraph_bidir(self):
+        g = treeGraphGen(2,2, kind='bidir')
+        self.assertEqual(list(g.edges()), [
+            (1, 0), (1, 3), (1, 4), (0, 1), (0, 2), (2, 0),
+            (2, 5), (2, 6), (3, 1), (4, 1), (5, 2), (6, 2)
+        ])
+
+# this is the way I might test that the types assigned in Z's functions are correct 
+# I think it would be a good idea to abstract out that logic into a helper function
```
Comment (teamdandelion) [inline infra/graph_generators_test.py:56]: yup, agreed! see my comment above

**Review by teamdandelion — COMMENTED** [review]
First pass review. Lots of bits of Python coding style to internalize. :)

File: infra/graph_generators.py (hunk 1 above)
Comment (mzargham) [reply infra/graph_generators.py:6]: Actually this was autogenerated by spyder and I just sort of ignored it. If you have a specific conversion for headers, I'll happily follow it.

Comment (teamdandelion) [reply infra/graph_generators.py:6]: Let's just have a standard python docstring describing the purpose of the module.

File: infra/graph_generators.py (hunk 2 above)
Comment (mzargham) [reply infra/graph_generators.py:10]: The the python versions of the pagerank algorithm use the types to look up the edge weights. In these simple generators there is only one type of edge but i was expecting to use them for constructing more sophisticated graphs by combining these, the resulting graphs would would have multiple types.

File: infra/graph_generators.py (hunk 8 above)
Comment (mzargham) [reply infra/graph_generators.py:45]: I can look into it; this was a work around in the first place. Some of the built in methods in networkx don't do quick what we want when initiated. Also, weird stuff happens when you iterate on the same graph you are editing.

**Comment by mzargham** [comment]
FYI @decentralion ~ i was out of sync with master and struggling to get lined up so to avoid issues we ran into with the other rebase, I opted to make a new branch. Working on these updates here:
https://github.com/sourcecred/research/tree/graph-generators-take-2/infra

File: infra/graph_generators.py (hunk 2 above)
Comment (mzargham) [reply infra/graph_generators.py:10]: ran black

**Comment by teamdandelion** [comment]
@mzargham cool, send a pull request my way once you want another review.

File: infra/graph_generators.py (hunk 2 above)
Comment (mzargham) [reply infra/graph_generators.py:10]: Added doc string; not sure they are totally up to convention but i read the link and added simple statements about each function.

File: infra/graph_generators_test.py (hunk 18 above)
Comment (mzargham) [reply infra/graph_generators_test.py:56]: factored it out. Thanks.

File: infra/graph_generators_test.py (hunk 15 above)
Comment (mzargham) [reply infra/graph_generators_test.py:24]: added error handling (first try)

File: infra/graph_generators.py (hunk 6 above)
Comment (wchargin) [reply infra/graph_generators.py:31]: Pardon the drive-by, but:

 1. Don’t use `is` (reference equality) when you want value equality:

    ```python
    >>> x = "a" * 1000
    >>> y = "a" * 1000
    >>> x == y
    True
    >>> x is y
    False
    ```

 2. Just use `in`:

    ```python
    if kind not in ("source", "sink", "bidir"):
        ...
    ```

Comment (teamdandelion) [reply infra/graph_generators.py:31]: Thanks for the correction. I think I tried the following test and concluded that `is` was ok:
```python
>>> a = "foo"
>>> b = "foo"
>>> a is b
True
```

I guess Python deduplicates literal strings but not constructed ones?

File: infra/graph_generators.py (hunk 2 above)
Comment (mzargham) [reply infra/graph_generators.py:10]: i updated these methods so the user can assign them with a dictionary, thus making heterogenous types, like in the sourcecred graphs with real data.

File: infra/graph_generators_test.py (hunk 15 above)
Comment (mzargham) [reply infra/graph_generators_test.py:24]: Added error handling across the module.

File: infra/graph_generators_test.py (hunk 17 above)
Comment (mzargham) [reply infra/graph_generators_test.py:36]: I kept them at mostly 3 and 4 because for 2 node graphs the there would be no way to distinguish some of the graph patterns; for example the star and the line would be the same.

**Comment by teamdandelion** [comment]
Closing as obsoleted by #26
